src/main/scala/shine/OpenCL/compilation/HostManagedBuffers.scala

```diff
@@ -0,0 +1,271 @@
+package shine.OpenCL.compilation
+
+import shine.DPIA._
+import shine.DPIA.Phrases._
+import shine.DPIA.Semantics.OperationalSemantics
+import shine.DPIA.Types._
+import shine.DPIA.primitives.{imperative => dpia}
+import shine.OpenCL._
+import shine.OpenCL.primitives.{imperative => ocl}
+
+import scala.annotation.tailrec
+import scala.collection.{immutable, mutable}
+
+object HostManagedBuffers {
+  def populate(params: immutable.Seq[Identifier[ExpType]], outParam: Identifier[AccType])
+  : Phrase[CommType] => Phrase[CommType] = { p =>
+    def outsideParams() = mutable.Set[Identifier[_ <: PhraseType]]() ++ (params :+ outParam)
+    val worstCasePrevious =
```

Comment:
Can you explain the terminology here? Why previous?

Reply:
`Metadata` represents effects, these are the effects that could have happened previously: before the host function was called. I can add a comment.

Reply:
Can we maybe try to replace the name `Metadata` as it is very generic and not helpful in the understanding of the code?
Maybe: `ReadsAndWrites`?

Reply:
I have renamed `Metadata` to `ReadsAndWrites`